fix flake8 config and drop unused code

wochinge opened the pull request:
**Proposed changes**:
- drop unused code
- re-configure flake8 to flag unused code / imports

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

m-vdb commented on rasa/core/channels/twilio.py:
```diff
@@ -10,7 +10,7 @@
 
 if TYPE_CHECKING:
-    from twilio.rest.api.v2010.account.message import MessageInstance
+    from twilio.rest.api.v2010.account.message import MessageInstance  # noqa: F401
```
sounds a bit weird to require this comment, can't vulture respect `TYPE_CHECKING`?

wochinge replied:
yeah - it sometimes works and sometimes doesn't (e.g. it doesn't raise for any other imports in `TYPE_CHECKING`)

m-vdb replied:
sounds a potential for flakiness. What can we do about it? Is this an issue Vulture people know about?

wochinge replied:
Funny enough: They use `rasa_nlu` to discuss the handling of `if TYPE_CHECKING` 😆

wochinge replied:
omg, our flake8 config is evil! Will probably dump vulture and use flake8 instead

m-vdb replied:
wdym?

wochinge replied:
we configured flake8 to not flag unused imports 🙈

m-vdb commented on rasa/core/featurizers/single_state_featurizer.py:
```diff
@@ -378,7 +378,7 @@ def __init__(self) -> None:
         super().__init__()
         rasa.shared.utils.io.raise_warning(
```
can you use `raise_deprecation_warning` instead?

wochinge replied:
sure, also changing it to `FutureWarning` . `DeprecationWarning` doesn't make sense here.

m-vdb replied:
why? I think that it has implication on how it is displayed to users. I wouldn't change this without looking at the impact on UX, and I think it's beyond the scope of the PR 🤔

wochinge replied:
That's exactly why. Users should see the warning and hence it should be `FutureWarning`. `DeprecationWarnings` don't print to the shell and make only sense if this is used from the Python API of rasa which isn't the case for these featurizers.

m-vdb commented on rasa/nlu/utils/__init__.py:
```diff
@@ -6,13 +6,6 @@
 import rasa.shared.utils.io
 
```
FWIW none of the remaining utils in the file seem NLU specific

wochinge replied:
yep, I'd say it's out of the scope of this PR though.

m-vdb commented on rasa/shared/core/events.py:
```diff
@@ -111,18 +111,6 @@ def format_message(
     )
 
 
-def first_key(d: Dict[Text, Any], default_key: Any) -> Any:
```
I guess you've checked that everything that is in `rasa.shared.*` and removed in this PR isn't used by Rasa X?

wochinge replied:
yes!

m-vdb commented on rasa/shared/data.py:
```diff
@@ -69,6 +70,10 @@ def get_test_directory(paths: Optional[Union[Text, List[Text]]]) -> Text:
     Returns:
         Path to temporary directory containing all found Core training files.
     """
+    rasa.shared.utils.io.raise_warning(
```
can you use `raise_deprecation_warning`?

wochinge replied:
changed these back as I need type `DeprecationWarning` in this case and I can't specify a category with `raise_deprecation_warning`

m-vdb replied:
why do you specifically need `DeprecationWarning`? What's the requirement from the user perspective?

wochinge replied:
I need a deprecation warning because I want to warn other developers or users using the Python API. This isn't called as part of anything when using the CLI (this would make a `FutureWarning` necessary). **If** anybody decides to use that in their Python code then they'll see that it's deprecated (e.g. Pycharm automatically strikes it through)

m-vdb commented on rasa/shared/importers/autoconfig.py:
```diff
@@ -263,7 +263,7 @@ def _get_commented_out_autoconfig_lines(
 
     for key in auto_configured_keys:
-        stream = yaml.compat.StringIO()
+        stream = ruamel.yaml.compat.StringIO()
```
but `ruamel.yaml` is imported as `yaml` above 🤷🏻

wochinge replied:
yeah 🤷🏻 I somewhat agree with vulture that it is weird to import a module / package by alias and then access a submodule in this package through the alias.

m-vdb replied:
can we either remove the renaming of the module or tweak vulture to disregard these? sounds like a bug of vulture

wochinge replied:
done

m-vdb commented on rasa/shared/nlu/training_data/message.py:
```diff
@@ -179,11 +179,10 @@ def get_full_intent(self) -> Text:
-        """Get intent as it appears in training data"""
-
+        """Get intent as it appears in training data."""
         rasa.shared.utils.io.raise_warning(
```
can you use `raise_deprecation_warning`?

m-vdb commented on rasa/shared/nlu/training_data/training_data.py:
```diff
@@ -464,8 +464,12 @@ def sorted_entities(self) -> List[Any]:
-        """Sorts the intent examples by the name of the intent and then response"""
-
+        """Sorts the intent examples by the name of the intent and then response."""
+        rasa.shared.utils.io.raise_warning(
```
+1

m-vdb commented on tests/nlu/test_persistor.py:
```diff
@@ -7,51 +9,13 @@
     pass
 
 
-async def test_list_method_method_in_AWS_persistor(  # noinspection PyPep8Naming
```
why removing all of these tests?

wochinge replied:
I removed the list methods of all persistors as it's not used anywhere. Hence, the tests are going with it.

m-vdb reviewed (commented):
Do you have an example of how it runs in the CI? I can't find an example of workflow with a failing check. Do we need some sort of documentation to explain what to do if the check fails?

wochinge commented:
pushed a commit which triggers a vulture error. Also happy to skip adding vulture to the CI for now. It's not perfect but in my opinion a good start to have a least some check for it.

m-vdb commented:
@wochinge it didn't fail because the lint for docstrings failed first :D

I don't think it's a good idea to not add it to the CI, because it will become dead code too (a bit ironic ahah)

wochinge commented on setup.cfg:
```diff
@@ -22,7 +22,10 @@ max-doc-length = 88
+ignore = W503, # line break before binary operator
+    E203,  # whitespace before ‘:’
+    E231,  # missing whitespace after ‘,’, ‘;’, or ‘:’
+    E501,  # too long lines - FIX THIS IN SEPARATE PR
```
will tackle this separately

wochinge replied:
https://github.com/RasaHQ/rasa/issues/8490

wochinge commented:
I dropped vulture and now use flake8 instead @m-vdb Removed more unused imports - flake8 is a lot better than vulture for this. Dropped vulture since it has to many false positives.

m-vdb commented:
but can flake8 detect dead code? unused import are a subset of dead code

wochinge commented:
not as good as vulture - but vulture has to many false positives so we can't really run it as part of the CI

wochinge commented:
@m-vdb Would be great if we could get this in before my vacation 👀 The review should be quick despite its size as it's just removing dead code.

m-vdb commented on rasa/nlu/components.py:
```diff
@@ -13,7 +13,6 @@
 from rasa.shared.exceptions import InvalidConfigException
 from rasa.shared.nlu.training_data.training_data import TrainingData
 from rasa.shared.nlu.training_data.message import Message
-from rasa.nlu.constants import COMPONENT_INDEX, TRAINABLE_EXTRACTORS
```
`COMPONENT_INDEX` seems to be needed at line 488

wochinge commented on setup.cfg:
```diff
@@ -22,7 +22,10 @@ max-doc-length = 88
-ignore = W503, E121, E126, E211, E225, E501, E203, E402, F401, F811, E231,
+ignore = W503, # line break before binary operator
+    E203,  # whitespace before ‘:’
+    E231,  # missing whitespace after ‘,’, ‘;’, or ‘:’
```
it's due to black - adding a comment for it

wochinge replied:
it's due to black - adding a comment for it

wochinge commented on rasa/cli/export.py:
```diff
@@ -12,10 +12,10 @@
 from rasa.shared.constants import DOCS_URL_EVENT_BROKERS, DOCS_URL_TRACKER_STORES
 from rasa.exceptions import PublishingError
 from rasa.shared.exceptions import RasaException
+from rasa.core.brokers.pika import PikaEventBroker
```
nope double checked. In my opinion we should aim for top level import where possible and just make exceptions for `tensorflow` (which is covered by a test)

wochinge commented on rasa/cli/run.py:
```diff
@@ -6,7 +6,7 @@
 from rasa.cli.arguments import run as arguments
 import rasa.cli.utils
-import rasa.shared.utils.cli
+import rasa.shared.utils.cli  # noqa: F401
```
I know - seems to be a false positive. Same as in `visualize - haven't found a way around it unfortunately.

m-vdb reviewed (approved):
⭐ thanks for the quick turnaround!
